core/store/src/lib.rs

```diff
@@ -187,9 +190,34 @@ impl StoreUpdate {
     }
 
     pub fn commit(self) -> Result<(), io::Error> {
-        if let Some(trie) = self.trie {
+        debug_assert!(
+            self.transaction.ops.len()
+                == self
+                    .transaction
+                    .ops
+                    .iter()
+                    .map(|op| match op {
+                        DBOp::Insert { col, key, .. } => {
+                            (*col as u8, key)
+                        }
+                        DBOp::Delete { col, key } => {
+                            (*col as u8, key)
+                        }
+                    })
+                    .collect::<std::collections::HashSet<_>>()
+                    .len(),
+            "Transaction overwrites itself: {:?}",
+            self
+        );
+        if let Some(trie) = self.tries {
             assert_eq!(
-                trie.storage.as_caching_storage().unwrap().store.storage.as_ref() as *const _,
+                trie.get_trie_for_shard(0)
```

Comment:
Can we use tries.get_store() here? That would be much easier to reason about otherwise, I would love to see a comment saying that the storage is the same for all the tries.